### PR TITLE
Config passporst.js with jwt & register & login endpoints

### LINKED ISSUE #6: Implement Basic Authentication with Passport.js and JWT
**Description:**
As part of our application's authentication system, we need to implement basic authentication using Passport.js and JWT (JSON Web Tokens). This will allow us to securely authenticate users and manage their sessions using JWT tokens.

**Tasks:**
1. Set up Passport.js and JWT dependencies in the project.
2. Configure Passport.js with the JWT strategy to handle authentication.
3. Implement user registration endpoint to create new user accounts securely.
4. Create a login endpoint that authenticates users and generates JWT tokens.
5. Implement a protected endpoint that requires JWT token for access.
6. Set up middleware to authenticate requests using Passport.js and the JWT strategy.
7. Handle logout functionality to invalidate tokens or sessions.
8. Implement password reset functionality for users who have forgotten their password.

**Acceptance Criteria:**
- Users should be able to register and create new accounts securely.
- Users should be able to log in and receive a JWT token for subsequent authenticated requests.
- The protected endpoint should only be accessible with a valid JWT token.
- JWT tokens should be properly validated, and expired or invalid tokens should be rejected.
- Users should be able to log out and invalidate their session or token.
- The password reset functionality should allow users to reset their password securely.

**Additional Information:**
- Make sure to follow best practices for security, such as properly hashing passwords and securely handling JWT tokens.
- Document the authentication API endpoints and any required input parameters or response formats.
- Ensure that appropriate error handling is implemented for various scenarios, such as invalid credentials or expired tokens.

**Related Dependencies or References:**
- Passport.js: [link to documentation](https://www.passportjs.org/)
- JWT: [link to documentation](https://jwt.io/)

### PR DESCRIPTION
**Pull request by sathsarabandaraj**
## Purpose

- The purpose of this PR is to fix #6 

## Goals

- Basic authentication to the API

## Approach

- Config passport.js Passport.js and JWT
- Add basic auth endpoints with controllers

##  Checklist
- [x] This PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.
- [x] I have read and understood the development best practices guidelines ( http://bit.ly/sef-best-practices )
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: src/configs/passport.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
+import passport from 'passport';
+import { Strategy as JwtStrategy, ExtractJwt } from 'passport-jwt';
+import { dataSource } from './dbConfig';
+import User from '../entity/user.entity';
+
+const JWT_SECRET = 'Nc$ykT*lUW3567^$@&iC1tpR$H$b29HJ'; // Same secret used for token creation
```
Comment (anjula-sack) [inline src/configs/passport.ts:6]: Use the env variable here

File: src/entity/user.entity.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
+// src/entity/user.entity.ts
```
Comment (anjula-sack) [inline src/entity/user.entity.ts:1]:
```suggestion
```

**Review by anjula-sack — CHANGES_REQUESTED** [review]
Good job! @sathsarabandaraj 
Can you the relevant tests to this?

File: src/app.ts [hunk 3]
```diff
@@ -2,15 +2,21 @@ import express from 'express';
...
 const app = express();
 app.use(bodyParser.json());
 app.use(cors());
+app.use(passport.initialize()); 
 
 app.get('/', (req, res) => {
   res.send('ScholarX Backend');
 });
 
+app.use('/api', userRouter);
```
Comment (anjula-sack) [inline src/app.ts:20]: Let's change them like below, 

```suggestion
app.use('/api/auth', authRouter);
```

**Review by anjula-sack — CHANGES_REQUESTED** [review]

File: src/controllers/user.controller.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+    const hashedPassword = await bcrypt.hash(password, 10);
+    const newUser = userRepository.create({ email, password: hashedPassword, contact_email, first_name, last_name, image_url, linkedin_url });
+    await userRepository.save(newUser);
+
+    res.status(201).json(newUser);
+  } catch (err) {
+    console.error('Error executing query', err);
+    res.status(500).json({ error: 'Internal server error' });
+  }
```
Comment (anjula-sack) [inline src/controllers/user.controller.ts:38]: Let's move this to services, authServices.ts

**Review by anjula-sack — CHANGES_REQUESTED** [review]

File: src/controllers/user.controller.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+dotenv.config();
+
+const JWT_SECRET = process.env.JWT_SECRET!;
+
+export const register = async (req: Request, res: Response) => {
+  try {
+    const { email, password, contact_email, first_name, last_name, image_url, linkedin_url } = req.body;
+
+    if (!email || !password || !contact_email || !first_name || !last_name || !image_url || !linkedin_url) {
```
Comment (anjula-sack) [inline src/controllers/user.controller.ts:17]: Only these are required. @sathsarabandaraj, set null to the other values

```suggestion
    if (!email || !password) {
```

**Review by anjula-sack — CHANGES_REQUESTED** [review]

File: src/app.ts [hunk 6]
```diff
@@ -2,15 +2,21 @@ import express from 'express';
...
 const app = express();
 app.use(bodyParser.json());
 app.use(cors());
+app.use(passport.initialize()); 
 
 app.get('/', (req, res) => {
   res.send('ScholarX Backend');
 });
 
+app.use('/api/auth', userRouter);
```
Comment (anjula-sack) [inline src/app.ts:20]:
```suggestion
app.use('/api/auth', authRoutes);
```

File: src/entity/user.entity.ts [hunk 7]
```diff
@@ -0,0 +1,53 @@
+import bcrypt from 'bcrypt';
```
Comment (anjula-sack) [inline src/entity/user.entity.ts:1]: let's rename this to `profile.entity.ts`

**Review by anjula-sack — CHANGES_REQUESTED** [review]

File: src/services/auth.service.ts [hunk 8]
```diff
@@ -0,0 +1,72 @@
+import { dataSource } from '../configs/dbConfig';
+import { Request, Response } from 'express';
+import jwt from 'jsonwebtoken';
+import bcrypt from 'bcrypt';
+import User from '../entity/user.entity';
+import { FindOneOptions } from 'typeorm';
+import dotenv from 'dotenv';
+
+dotenv.config();
```
Comment (anjula-sack) [inline src/services/auth.service.ts:9]: We don't have to import this since we are using it on app.ts
```suggestion
```

Comment (anjula-sack) [reply src/services/auth.service.ts:9]: Is there an issue without it?

File: src/configs/passport.ts [hunk 9]
```diff
@@ -0,0 +1,31 @@
+import passport from 'passport';
+import { Strategy as JwtStrategy, ExtractJwt } from 'passport-jwt';
+import { dataSource } from './dbConfig';
+import User from '../entity/user.entity';
+import dotenv from 'dotenv';
+
+dotenv.config();
```
Comment (anjula-sack) [inline src/configs/passport.ts:7]:
```suggestion
```

**Review by anjula-sack — CHANGES_REQUESTED** [review]

File: src/app.ts [hunk 10]
```diff
@@ -2,15 +2,27 @@ import express from 'express';
...
 });
 
-export default app;
+app.use('/api/auth', authRouter);
+
+const server = app.listen(port, () => {
+  console.log(`Server is running on http://localhost:${port}`);
+});
+
+export default server;
```
Comment (anjula-sack) [inline src/app.ts:28]:
```suggestion
export default server;

```

Add the end of line

File: src/entity/profile.entity.ts [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+  last_name: string;
+
+  @Column({ type: 'varchar', length: 255 })
+  image_url: string;
+
+  @Column({ type: 'varchar', length: 255 })
+  linkedin_url: string;
+
+  @Column({ type: 'varchar', length: 10, default:"default"})
+  type: string;
```
Comment (anjula-sack) [inline src/entity/profile.entity.ts:29]: This should be an enum @sathsarabandaraj

File: src/entity/profile.entity.ts [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+
+  @BeforeInsert()
+  async generateUuid() {
+    if (!this.uuid) {
+      this.uuid = uuidv4();
+    }
+  }
+}
+
+export default Profile;
```
Comment (anjula-sack) [inline src/entity/profile.entity.ts:81]:
```suggestion
export default Profile;

```

**Review by anjula-sack — CHANGES_REQUESTED** [review]
Requested small changes

**Review by anjula-sack — APPROVED** [review]
Great job! @sathsarabandaraj

**Comment by Piumal1999** [comment]
Hi @sathsarabandaraj, Could you please update the readme file with the instructions for setting the JWT secret key?

**Pull request merged**